sql/ir_generator_test.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019 The SQLFlow Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"testing"
+
+	"github.com/sql-machine-learning/sqlflow/sql/codegen"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGenerateTrainIR(t *testing.T) {
+	a := assert.New(t)
+	parser := newParser()
+
+	normal := `SELECT c1, c2, c3,c4
+FROM my_table
+TRAIN DNNClassifier
+WITH model.n_classes=2, train.optimizer="adam"
+COLUMN c1,NUMERIC(c2, [128, 32]),CATEGORY_ID(c3, 512),
+       SEQ_CATEGORY_ID(c3, 512),
+	   CROSS([c1,c2], 64),
+	   BUCKET(NUMERIC(c1, [100]), 100),
+	   EMBEDDING(CATEGORY_ID(c3, 512), 128, mean)
+LABEL c4
+INTO mymodel;`
+
+	r, e := parser.Parse(normal)
+	a.NoError(e)
+
+	trainIR, err := generateTrainIR(r, "mysql://somestring")
+	a.NoError(err)
+	a.Equal("DNNClassifier", trainIR.Estimator)
+	a.Equal("SELECT c1, c2, c3, c4\nFROM my_table", trainIR.Select)
+
+	// TODO(typhoonzero): should find out why the order of parsed attributes changes randomly
```

Comment:
Good catch. This is due the `attrs` type in the parser is a map.

https://github.com/sql-machine-learning/sqlflow/blob/01730675a6505d1384393a04f647f2dc380a211e/sql/sql.y#L87

During the parsing, the key-value pairs of `attrs` is assigned as

https://github.com/sql-machine-learning/sqlflow/blob/01730675a6505d1384393a04f647f2dc380a211e/sql/sql.y#L260-L267

I find it hard to switch from map to slice because multiple code generators are depending on the map struct. Also, there is a check on the duplicating map keys in our parser.

https://github.com/sql-machine-learning/sqlflow/blob/01730675a6505d1384393a04f647f2dc380a211e/sql/sql.y#L103-L111

@weiguoz @typhoonzero @Yancey1989 Sorry I didn't realize this issue. Shall we change the type of `TrainIR.Attribute` back to `map[string]interface{}`?

Reply:
Agree that we should use map for now.